crates/typst/src/visualize/pattern.rs

```diff
@@ -0,0 +1,241 @@
+/// `{"parent"}`. To create word-by-word or glyph-by-glyph patterns, you can
+/// wrap the words or characters of your text in [boxes]($box) manually or
+/// through a [show rule]($styling/#show-rules).
+///
+/// ```example
+/// #let pat = pattern(
+///   (30pt, 30pt),
+///   relative: "parent",
+///   square(size: 30pt, fill: gradient.conic(..color.map.rainbow))
+/// );
+///  #set text(fill: pat)
+///  #lorem(10)
+/// ```
+///
+/// You can also space the elements further or closer apart using the
+/// [`spacing`]($pattern.spacing) feature of the pattern. If the spacing
+/// is lower than the size of the pattern, the pattern will overlap.
+/// If it is higher, the pattern will have gaps of the same color as the
+/// background of the pattern.
+///
+/// ```example
+/// #let pat = pattern(
+///   (30pt, 30pt),
+///   spacing: (10pt, 10pt),
+///   relative: "parent",
+///   square(size: 30pt, fill: gradient.conic(..color.map.rainbow))
+/// );
+///  #rect(width: 100%, height: 100%, fill: pat)
+/// ```
+///
+/// # RelativeToness
+/// The location of the starting point of the pattern is dependant on the
+/// dimensions of a container. This container can either be the shape they
+/// are painted on, or the closest surrounding container. This is controlled by
+/// the `relative` argument of a pattern constructor. By default, patterns are
+/// relative to the shape they are painted on, unless the pattern is applied on
+/// text, in which case they are relative to the closest ancestor container.
+///
+/// Typst determines the ancestor container as follows:
+/// - For shapes that are placed at the root/top level of the document, the
+///   closest ancestor is the page itself.
+/// - For other shapes, the ancestor is the innermost [`block`]($block) or
+///   [`box`]($box) that contains the shape. This includes the boxes and blocks
+///   that are implicitly created by show rules and elements. For example, a
+///   [`rotate`]($rotate) will not affect the parent of a gradient, but a
+///   [`grid`]($grid) will.
+#[ty(scope)]
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub struct Pattern {
+    /// The body of the pattern
+    pub body: Prehashed<Content>,
+    /// The pattern's rendered content.
+    pub frame: Prehashed<Frame>,
+    /// The pattern's tile size.
+    pub bbox: Size,
+    /// The pattern's tile spacing.
+    pub spacing: Size,
+    /// The pattern's relative transform.
+    pub relative: Smart<RelativeTo>,
+}
+
+impl Eq for Pattern {}
+
+#[scope]
+impl Pattern {
+    /// Construct a new pattern.
+    ///
+    /// ```example
+    /// #let pat = pattern(
+    ///   (20pt, 20pt),
+    ///   relative: "parent",
+    ///   place(dx: 5pt, dy: 5pt, rotate(45deg, square(size: 5pt, fill: black)))
+    /// );
+    ///  #rect(width: 100%, height: 100%, fill: pat)
+    /// ```
+    #[func(constructor)]
+    pub fn construct(
+        vm: &mut Vm,
+        /// The bounding box of each cell of the pattern.
+        bbox: Spanned<Axes<Length>>,
```

Comment:
isn't `size` a more fitting name? you could also implement `Cast` for `Abs` perhaps to skip on the spanned.

Reply:
also, could the pattern size theoretically be `auto`, since you're layouting anyway?